engine/common/follower/cache/cache.go

```diff
@@ -0,0 +1,144 @@
+package cache
+
+import (
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	herocache "github.com/onflow/flow-go/module/mempool/herocache/backdata"
+	"github.com/onflow/flow-go/module/mempool/herocache/backdata/heropool"
+)
+
+type OnEquivocation func(first *flow.Block, other *flow.Block)
+
+// Cache stores pending blocks received from other replicas, caches blocks by blockID, it also
+// maintains secondary index by view and by parent. Additional indexes are used to track proposal equivocation(multiple
+// valid proposals for same block) and find blocks not only by parent but also by child.
+// Resolves certified blocks when processing incoming batches.
+// Concurrency safe.
+type Cache struct {
+	backend *herocache.Cache // cache with random ejection
+	lock    sync.RWMutex
+	// secondary index by view, can be used to detect equivocation
+	byView map[uint64]*flow.Block
+	// secondary index by parentID, can be used to find child of the block
+	byParent map[flow.Identifier]*flow.Block
+	// when message equivocation has been detected report it using this callback
+	onEquivocation OnEquivocation
+}
+
+// Peek performs lookup of cached block by blockID.
+// Concurrency safe
+func (c *Cache) Peek(blockID flow.Identifier) *flow.Block {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if block, found := c.backend.ByID(blockID); found {
+		return block.(*flow.Block)
+	} else {
+		return nil
+	}
+}
+
+// NewCache creates new instance of Cache, as part of construction process connects ejection event from HeroCache to
+// post-ejection processing logic to perform cleanup of secondary indexes to prevent memory leaks.
+func NewCache(log zerolog.Logger, limit uint32, collector module.HeroCacheMetrics, onEquivocation OnEquivocation) *Cache {
+	distributor := NewDistributor(collector)
+	cache := &Cache{
+		backend: herocache.NewCache(
+			limit,
+			herocache.DefaultOversizeFactor,
+			heropool.RandomEjection,
+			log.With().Str("follower", "cache").Logger(),
+			distributor,
+		),
+		byView:         make(map[uint64]*flow.Block, 0),
+		byParent:       make(map[flow.Identifier]*flow.Block, 0),
+		onEquivocation: onEquivocation,
+	}
+	distributor.AddConsumer(cache.handleEjectedEntity)
+	return cache
+}
+
+// handleEjectedEntity performs cleanup of secondary indexes to prevent memory leaks.
+// WARNING: Concurrency safety of this function is guaranteed by s.lock, this callback can be called
+// only in herocache.Cache.Add and we perform this call while s.lock is in locked state.
+func (c *Cache) handleEjectedEntity(entity flow.Entity) {
+	block := entity.(*flow.Block)
+	delete(c.byView, block.Header.View)
+	delete(c.byParent, block.Header.ParentID)
```

Comment:
Suppose I have a chain like:
```
A <- B
  ^- C
```

I insert B, then C, then A. But between inserting C and A, B is ejected. When B is ejected, `B.ParentID == A.ID()` is removed from `byParent`. Then I will have a valid child for A in the cache, but not indexed, when I insert A, meaning that I won't report A as certified. 

Depending on the access pattern to the cache, this could cause liveness issues. For example, suppose the caller checks whether a block already exists in the cache before adding it (using `Peek`). Extending the example above, further suppose that B is orphaned and we never receive it again, while further progress is made building on C. We may receive C again from the sync engine, but the caller might observe that it is already in the cache and drop it.

Reply:
This shouldn't happen since we add to the cache before applying change to the index. 
We would get:
Try to insert B
B inserted
index state: `byParent = { B.ParentID: A.ID() }`

Try to insert C
B gets ejected, index state: `byParent = {}`
C inserted
index state: `byParent = { C.ParentID: A.ID() }`

Try to insert A
A inserted
index state: `byParent = { C.ParentID: A.ID(), A.ParentID: R.ID() }`

In this scenario it has to be resolved.
Moreover we expect that sync engine will deliver blocks again and again as long as they are not finalized.

Reply:
In your example, B is ejected while inserting C. The scenario I am thinking of is if B is ejected while inserting some other unrelated block Z, which is unconnected to A, B, or C, after C and before A. Then we would have:

Try to insert B
B inserted
index state: `byParent = { B.ParentID: A.ID() }`

Try to insert C
C inserted
index state: `byParent = { B.ParentID: A.ID() }`

Try to insert Z
Z inserted
B gets ejected, index state: `byParent = { <Z parent> }`

Try to insert A
A inserted
index state: `byParent = { <Z parent>, <A parent>}`

> Moreover we expect that sync engine will deliver blocks again and again as long as they are not finalized.

I agree, however we are additionally relying on the Follower engine to _not_ de-duplicate those blocks using the cache for correctness, because of this edge case.

Reply:
Ok, that makes sense, I think we need to change the index from `map[flow.Identifier]*flow.Block` to `map[flow.Identifier]map[flow.Identifier]*flow.Block`, this way we will be able to track all children with same parent.

Reply:
@jordanschalm I've reworked `byParent` index and implemented additional tests: https://github.com/onflow/flow-go/pull/3965/commits/ecd9e9464a1eb20733d08f606beb3a964a43a4e3

Reply:
Thanks for adding the test cases 🙇